### ISSUE TITLE
strange values when reset no3 and nh4 in a fixed date

### ISSUE THREAD

**Issue by JJguri** [APSIMInitiative/APSIMClassic#2080]
I am trying to reset the n03 and nh4 values for a given date where I have soil sampling information. I modified the reset manager rule which is by default. However, I am not getting the right values I want. Here is an example: I know that for this given date I had 161.7 ppm of no3 and 198.9 ppm of nh4, however if I distribute these amounts across the profile (numbers in the script below) the reset values are 79.253 ppm of no3() and 68.278 nh4(). I am reporting daily `no3()` and `nh4()` assuming it is the sum of all layers in ppm.

Here is the script:

```

         if (today = date('[reset_date]')) then
            resetWater = '[resetWater]'
            resetNitrogen  = '[resetNitrogen]'
            resetSurfaceOM = '[resetSurfaceOM]'
            resetP = '[resetP]'
            if (resetWater = 'yes') then
                '[soilmodule] Water' reset
            endif
            if (resetNitrogen = 'yes') then
                '[soilmodule] Nitrogen' set no3 = 26.71 15 15 15 15 15 15 15 15 15
                '[soilmodule] Nitrogen' set nh4 = 20 20 20 20 20 20 20 20 20 18.9
            endif
            if (resetSurfaceOM = 'yes') then
                '[surfaceommodule]' reset
            endif
            if (resetP = 'yes') then
                '[pmodule]' reset
            endif
            act_mods reseting
         endif
```

and here the output:
![image](https://user-images.githubusercontent.com/41041705/128114342-cff7613d-4c52-425c-9e96-a9138144b6f1.png)

**Comment by BrianCollinss** [comment]
Are you resetting at the end of the day or beginning?

**Comment by JJguri** [comment]
start_of_day

**Comment by BrianCollinss** [comment]
I am not sure, but it could be the reason. Could you try doing it at the end of the orevious day?

**Comment by BrianCollinss** [comment]
When you reset water, it just recerst it back to the number you have in InitialWater. So you can set that number 50% instead and just reset water as it is done by default.

**Comment by JJguri** [comment]
> When you reset water, it just recerst it back to the number you have in InitialWater. So you can set that number 50% instead and just reset water as it is done by default.

yes that is a good option, I am doing that now, however, the measurement I have is the amount of soil moisture in grams per hectograms, for example 10.485, so I am wondering which would be the best way to transform this value to percentage of water available to put in apsim in the initial water node.

**Comment by BrianCollinss** [comment]
I think you can use BD to convert the values.

**Comment by JJguri** [comment]
> I am not sure, but it could be the reason. Could you try doing it at the end of the orevious day?

I re-run using end_of_day and applying the reset the previous day, I got the following values: 162.508 (no3ppm()), 197.326 (nh4ppp()) using the following code:

```
         if (today = date('[reset_date]')) then
            resetWater = '[resetWater]'
            resetNitrogen  = '[resetNitrogen]'
            resetSurfaceOM = '[resetSurfaceOM]'
            resetP = '[resetP]'
            if (resetWater = 'yes') then
                '[soilmodule] Water' reset
            endif
            if (resetNitrogen = 'yes') then
                '[soilmodule] Nitrogen' set no3ppm = 23 15 15 15 15 15 15 15 5 0 (mg/kg)
                '[soilmodule] Nitrogen' set nh4ppm = 49.5 20 20 20 20 20 20 20 20 18.9 (mg/kg)
            endif
            if (resetSurfaceOM = 'yes') then
                '[surfaceommodule]' reset
            endif
            if (resetP = 'yes') then
                '[pmodule]' reset
            endif
            act_mods reseting
         endif
```

**Comment by BrianCollinss** [comment]
So it is closer now! Look at the values of the previous day if you report at the end of the day.

**Comment by peter-devoil** [comment]
setting sw_dep, not esw, might work.

**Comment by JJguri** [comment]
> setting sw_dep, not esw, might work.

this is total water in the soil right? no related to LL15?

**Comment by peter-devoil** [comment]
Yes. You can find which variables are settable in the "set_my_variable" subroutine of [the code](https://github.com/APSIMInitiative/APSIMClassic/blob/cfeceaf62b2ca0435e7a9df9b93b68eee702ffcd/Model/SoilWat/SoilWat.for#L4492)

**Comment by JJguri** [comment]
Hi @peter-devoil I am trying to set the oc values with specific values for a given date and it is not working. This code is being applied at the `end_of_day` and with `yes` in the resetting tab on `properties` for Nitrogen:

```
         if (today = date('[reset_date]')) then
            resetWater = '[resetWater]'
            resetNitrogen  = '[resetNitrogen]'
            resetSurfaceOM = '[resetSurfaceOM]'
            resetP = '[resetP]'
            if (resetWater = 'yes') then
                '[soilmodule] Water' reset
            endif
            if (resetNitrogen = 'yes') then
                '[soilmodule] Nitrogen' set oc = 0.8 0.8 0.764 0.79 0.627 0.622 0.425 0.001 0 0 (%)
            endif
            if (resetP = 'yes') then
                '[pmodule]' reset
            endif
            act_mods reseting
         endif
```
![image](https://github.com/APSIMInitiative/APSIMClassic/assets/41041705/7022b94d-fbf0-498c-8093-6af9bb9cb4fc)

**Comment by peter-devoil** [comment]
Hi,

The variable name is right, but you see in that code that it gets all cranky-pants about setting the variable during a simulation (as well, you're linking to SoilNitrogen, not [SoilN](https://github.com/APSIMInitiative/APSIMClassic/blob/master/Model/SoilN/SoilN.f90#L1630). 

That's because oc is the sum of 2 om pools - Humic and Biomass - and the implementors were too ~~lazy~~ cautious to implement code that would partition the amount you want to set it to into sub pools - it's probably a good idea to ask why the OM pools get so wrong in the first place.

So you have to do it yourself. Calculate a ratio (or use the ratio you started the simulation with) of hum:biom; and then set the hum/biom pools for that particular concentration of OC. Good luck with units (ppm vs kg/ha). Code is [here](https://github.com/APSIMInitiative/APSIMClassic/blob/master/Model/SoilN/SoilN.f90#L5175).

**Comment by JJguri** [comment]
Hi @peter-devoil thanks a lot for the detailed explanation. 
- I checked the SoilN documentation and I found `carbon_tot` as variable. Is this the sum of `hum_c` and `biom_c`? I was thinking to convert oc (%) to carbon_tot (kg/ha) and reset this variable for a given date. 
- In terms of the soil module, do you mean I should use `'[soilmodule] SoilN'` set instead of `'[soilmodule] Nitrogen'` set?
- Regarding units shouldI used bulk density in each layer to convert % to kg/ha?

**Comment by peter-devoil** [comment]
- it's the sum of hum_c, biom_c and inert_c IIRC
- Yes it's confusing. Look at the top of the summary file to see the list of module names (LHS) and dll names (RHS). The dll name reflects the code folder in the source tree. The module name is what you put in the manager script - so you want `'[soilmodule] Nitrogen'`
- You do need BD to convert to ppm, so yes, you'll need it for %age as well

**Comment by JJguri** [comment]
@peter-devoil thanks for the explanations. Do you know where I can find lower and upper possible values for the parameters specified in Table 1 [here](https://www.apsim.info/documentation/model-documentation/soil-modules-documentation/soiln/). I found oc is quite sensitive to some of these parameters.

**Comment by peter-devoil** [comment]
WRT oc/fbiom/finert that you talk about above - there's a bit of info in [this document](https://www.apsim.info/wp-content/uploads/2019/10/Parameters-for-soil-water-Ver24.pdf), though not much on variability.. You can compare soils in the database to see if they differ (most do not).

Most of the parameters in that table 1 are back calculated from "well behaved" simulations where we know the starting and final pool composition, and the decay rates are "optimised" by whatever method is available.
